File: backend/src/test/java/com/localassistant/os/service/ActionServiceTest.java
package com.localassistant.os.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import com.localassistant.os.config.AssistantProperties;
import com.localassistant.os.store.StateStore;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

class ActionServiceTest {
    @TempDir
    Path root;

    private ActionService actions;
    private MemoryService memories;

    @BeforeEach
    void setUp() throws Exception {
        AssistantProperties properties = new AssistantProperties();
        properties.setDataDir(root.resolve("data"));
        properties.setWorkspaceRoot(root);
        StateStore store = new StateStore(properties);
        store.initialize();
        WorkspaceService workspace = new WorkspaceService(properties);
        memories = new MemoryService(store);
        actions = new ActionService(store, workspace, memories);
    }

    @Test
    void proposedWritesDoNothingUntilApproved() throws Exception {
        var action = actions.propose(
                "workspace_write_file",
                Map.of("path", "notes/approved.txt", "content", "approved content"),
                "Test approval.");

        assertThat(root.resolve("notes/approved.txt")).doesNotExist();
        assertThat(actions.approve(action.id()).status()).isEqualTo("completed");
        assertThat(Files.readString(root.resolve("notes/approved.txt"))).isEqualTo("approved content");
    }

    @Test
    void rejectedMemoryIsNotStored() throws Exception {
        var action = actions.propose(
                "remember",
                Map.of("content", "Temporary", "category", "preference"),
                "Test rejection.");
        actions.reject(action.id());
        assertThat(memories.list()).isEmpty();
    }

    @Test
    void nonWebUrlsAreRejectedBeforeQueueing() {
        assertThatThrownBy(() -> actions.propose(
                "browser_open_url",
                Map.of("url", "file:///C:/Windows/System32/calc.exe"),
                "Should not queue."))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("Only HTTP and HTTPS");
    }
}

File: tests/action-service.test.ts
import assert from "node:assert/strict";
import { mkdtemp, readFile } from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import test from "node:test";
import { ActionService } from "../src/services/action-service.js";
import { MemoryService } from "../src/services/memory-service.js";
import { WorkspaceService } from "../src/services/workspace-service.js";
import { StateStore } from "../src/store/state-store.js";

async function createServices() {
  const root = await mkdtemp(path.join(os.tmpdir(), "assistant-actions-"));
  const store = new StateStore(path.join(root, "data"));
  await store.initialize();
  const workspace = new WorkspaceService(root);
  const memories = new MemoryService(store);
  const actions = new ActionService(store, workspace, memories);
  return { actions, memories, root };
}

test("a proposed file write does nothing until approved", async () => {
  const { actions, root } = await createServices();
  const action = await actions.propose(
    "workspace_write_file",
    { path: "notes/approved.txt", content: "approved content" },
    "Test the approval boundary.",
  );

  await assert.rejects(readFile(path.join(root, "notes", "approved.txt")));
  const completed = await actions.approve(action.id);

  assert.equal(completed.status, "completed");
  assert.equal(
    await readFile(path.join(root, "notes", "approved.txt"), "utf8"),
    "approved content",
  );
});

test("rejected memory proposals are not stored", async () => {
  const { actions, memories } = await createServices();
  const action = await actions.propose(
    "remember",
    { content: "A temporary preference", category: "preference" },
    "Test rejection.",
  );

  await actions.reject(action.id);
  assert.deepEqual(memories.list(), []);
});

test("non-web browser URLs are rejected before entering the queue", async () => {
  const { actions } = await createServices();
  await assert.rejects(
    actions.propose(
      "browser_open_url",
      { url: "file:///C:/Windows/System32/calc.exe" },
      "Should never be queued.",
    ),
    /Only HTTP and HTTPS URLs/,
  );
});
